File: windows/Scanner/Detectors/DetectorSupport.cs
using AIExposureScanner.Scanner.Filesystem;
using AIExposureScanner.Scanner.Models;

namespace AIExposureScanner.Scanner.Detectors;

internal static class DetectorSupport
{
    public static string HomePath(IFilesystemFacade fs, string suffix) =>
        string.IsNullOrEmpty(suffix)
            ? fs.HomeDirectory
            : Path.Combine(new[] { fs.HomeDirectory }.Concat(suffix.Split(['/', '\\'])).ToArray());

    public static void AppendConfigFileFact(ScanFacts facts, IFilesystemFacade fs, string appId, string path)
    {
        if (fs.FileExists(path))
        {
            facts.ConfigFiles.Add(new ConfigFileFact(appId, path, fs.IsWorldReadableFile(path)));
        }
    }

    public static void AppendAppInstallationFact(
        ScanFacts facts,
        IFilesystemFacade fs,
        string appId,
        IEnumerable<string> candidatePaths
    )
    {
        var evidencePath = candidatePaths.FirstOrDefault(fs.DirectoryExists);
        facts.AppInstallations.Add(
            evidencePath is null
                ? new AppInstallationFact(appId, false)
                : new AppInstallationFact(appId, true, evidencePath)
        );
    }

    public static string? ReadIfPresent(IFilesystemFacade fs, string path) =>
        fs.FileExists(path) ? fs.ReadTextFile(path) : null;
}

File: macos/Scanner/Sources/Scanner/Detectors/DetectorSupport.swift
func homePath(_ fs: any FilesystemFacade, _ suffix: String) -> String {
    if suffix.isEmpty {
        return fs.homeDirectory
    }
    return fs.homeDirectory + "/" + suffix
}

func appendConfigFileFact(_ facts: inout ScanFacts, fs: any FilesystemFacade, appId: String, path: String) {
    if fs.fileExists(path) {
        facts.configFiles.append(
            ConfigFileFact(appId: appId, path: path, worldReadable: fs.isWorldReadableFile(path))
        )
    }
}

func appendAppInstallationFact(
    _ facts: inout ScanFacts,
    fs: any FilesystemFacade,
    appId: String,
    candidatePaths: [String]
) {
    if let evidencePath = candidatePaths.first(where: { fs.directoryExists($0) }) {
        facts.appInstallations.append(AppInstallationFact(appId: appId, installed: true, evidencePath: evidencePath))
    } else {
        facts.appInstallations.append(AppInstallationFact(appId: appId, installed: false))
    }
}

func readIfPresent(_ fs: any FilesystemFacade, path: String) throws -> String? {
    guard fs.fileExists(path) else {
        return nil
    }
    return try fs.readTextFile(path, maxBytes: 10 * 1024 * 1024)
}
